### PR TITLE
Decouple from router (only build routes), simplify route processing, add lifecycle events, add redirects

### LINKED ISSUE #3: Break out router from lib, just use this lib for route rendering

### PR DESCRIPTION
**Pull request by wswoodruff**
### Issues I was running into
I was having issues with the <Switch> component and which routes were supposed to be rendering but weren't because Switch will render the first route that matches and none else.

Ex: `{ path: '/app/:someVar' }` will render if it appears in a list above  `{ path: '/app/path' }` and '/app/path' will never render.

Also - If you've got child routes, and your route doesn't match then none of your child routes will either
So if you have `{ path: '/app', exact: true, childRoutes: [{ path: 'subRoute' }] }`
Then you can never navigate to the component at `/app/subRoute`

- To remedy this, I you can have structural route parts that don't have any component like here: https://github.com/BigRoomStudios/nearpeer-frontend/blob/955e042957a4f2ef96c27cdc38bf1ccf6b6c4459/src/routes/app/index.js#L158-L160

### Some improvements provided in this PR
- Decouple the use of any routers inside this lib, only use it to render routes, given a config
- Remove the server-side rendering bit until we have a chance to test it out
- Introduce the `redirect` property that will pass all of its properties to a Redirect component

### Tips
- It helps to organize your paths by having a structural route part like linked above. See in this example for the path `app/dms/` here: https://github.com/BigRoomStudios/nearpeer-frontend/blob/955e042957a4f2ef96c27cdc38bf1ccf6b6c4459/src/routes/app/index.js#L109-L120
- Referencing the above link, to match a route at the structural path property you create a slash route with `{ path: '/', exact: true }` -- making sure to have exact set to true

### CODE REVIEW

**Comment by devinivy** [comment]
Hey @wswoodruff could you summarize the problems you encountered and changes made here?  Maybe best to in person?  Also, it seems like there are two major changes in here: 1. lifecycle methods and 2. a big refactor to fix some bugs.  We might want to update the title/description to reflect that.

File: lib/index.js [hunk 1]
```diff
@@ -1,98 +1,273 @@
...
+        .sort(internals.sortRoutes(baseRoute.path));
+
+    return clone;
+};
+
+internals.rFlattenChildRoutes = (route) => {
+
+    // Clone because we're building an army &&
+    // Transform to an array so we can play
+    let clone = [].concat(Object.assign({}, route));
```
Comment (wswoodruff) [inline lib/index.js]: Change this to `let clone = [{ ...route }]));`

File: lib/index.js [hunk 2]
```diff
@@ -1,98 +1,149 @@
...
-        // These relative paths seem to require a slash in front
+internals.rRenderRoute = (basePath) => {
 
-        let path = route.path;
-        if (path[0] !== '/') {
-            path = `/${path}`;
-        }
+    return (route) => {
+
+        const updatedPath = String(`${basePath}/${route.path}`).replace(/\/+/g, '/'); // Remove duplicate slashes
```
Comment (devinivy) [inline lib/index.js]: What cases does this cover?  Is it just two: `basePath` is `/` or is not `/`?  If so, let's just handle those cases explicitly.

File: lib/index.js [hunk 3]
```diff
@@ -1,98 +1,149 @@
...
+        const RouteComponent = route.component || 'div';
```
[the remark this answers is not in the document]
Comment (wswoodruff) [reply lib/index.js]: Do fragments get the component lifecycle as well?

File: lib/index.js (hunk 2 above)
Comment (wswoodruff) [reply lib/index.js]: At the moment it's possible to configure a path with a trailing slash and a childRoute that begins with a slash

File: lib/index.js [hunk 4]
```diff
@@ -1,98 +1,149 @@
...
+        }
+    }
+
+    render() {
+
+        return this.props.children;
+    }
+}
+
+internals.sortRoutes = (a, b) => {
```
Comment (devinivy) [inline lib/index.js]: I'm concerned about completely removing the control of route ordering from the user.  What allowed the nearpeer config to work?

File: lib/index.js [hunk 5]
```diff
@@ -1,98 +1,149 @@
...
-                    </route.component>
-                )}
+                render={(props) => {
+
+                    return (
+                        <internals.routeComponentLifecycleWrapper {...props} route={route}>
+                            <RouteComponent {...props} route={route}>
+                                {
+                                    route.childRoutes &&
+                                    route.childRoutes.length !== 0 &&
```
Comment (devinivy) [inline lib/index.js]: Is there anything wrong with having an empty `<Switch/>`?

Comment (wswoodruff) [reply lib/index.js]: 🤔 Good Q -- I'm not sure but figured if we could avoid it might as well? :man_shrugging:

File: lib/index.js [hunk 6]
```diff
@@ -1,98 +1,149 @@
...
-        return (
-            <ConnectedRouter
-                history={history}>
-                {internals.renderRoutes(routes)}
-            </ConnectedRouter>
-        );
-    }
-};
+    const rootSlashChildren = rootSlashRoutes.map((r) =>  r.childRoutes ? r.childRoutes : [])
+        .reduce((collector, r) => collector.concat(r), []);
```
Comment (devinivy) [inline lib/index.js]: What does this do?  Flatten just the first level of child routes?

File: lib/index.js (hunk 4 above)
Comment (wswoodruff) [reply lib/index.js]: Sorting the routes for users I feel is like a hodge-podge-y thing -- without that happening for you it's pretty easy for routes to not match as you'd expect. and I _think_ there's only 1 right way to sort these routes when putting them in a <Switch />

File: lib/index.js (hunk 2 above)
Comment (devinivy) [reply lib/index.js]: Ah, that makes sense.  Really small peas, but it would be sweet to make that more clear at some point—regex seems like a big lift to express this rule/ability, and it allows some weird stuff like making a route `////`.

File: lib/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (devinivy) [reply lib/index.js]: Aren't you wrapping this in a component that has the lifecycle?

[the remark this answers is not in the document]
Comment (wswoodruff) [reply lib/index.js]: Oh yeah haha!

File: lib/index.js (hunk 4 above)
Comment (devinivy) [reply lib/index.js]: Aiming for the principle of least astonishment here.  There's no way to use regex routing without ambiguity as to which of two routes are more specific, which is why the order matters.  That said, I would be interested to know how it is handled with that old route config—do you know what made that work?

Comment (wswoodruff) [reply lib/index.js]: Yeah this is a tough thing, before adding the sorting I was relying on just my ordering in the config to get things to work correctly

File: lib/index.js (hunk 6 above)
Comment (wswoodruff) [reply lib/index.js]: Yep! We need to flatten the first level of childRoutes to kind of 'kick off' the matching scheme

File: lib/index.js (hunk 4 above)
Comment (devinivy) [reply lib/index.js]: Relying on the order is typical in any regex-based setup, e.g. express, and the ability to change the order is really important.  The question remains: what allowed the order to matter "less" in that old route config that you showed earlier?

File: lib/index.js (hunk 6 above)
Comment (devinivy) [reply lib/index.js]: Why is that?  Wont the top-level `/` match any route?

Comment (wswoodruff) [reply lib/index.js]: I was running into issues with the root slash route stealing the render away from the root-level catchall route which is just `/:myParam`

Comment (devinivy) [reply lib/index.js]: Ah, interesting.  Maybe leave a comment in there about that issue, if that's why the code is there.

File: lib/index.js (hunk 4 above)
Comment (wswoodruff) [reply lib/index.js]: Then it's my stubbornness with a regex-based routing system prolly, I feel like when I'm writing routes I shouldn't have to worry about the order my array of routes is in, just that if I create a childRoute that it will be rendered at that path -- I get the feeling that I should get over that?

Perhaps I can have the user pass options and have this automatic sorting happen for them if they wish it to.

Comment (devinivy) [reply lib/index.js]: Yeah, this is why hapi wrote its own library for routing with a _more restricted_ matching functionality :) So that order doesn't matter.  But it just does matter with react-router, express, etc.  In most routing schemes it matters (because there simply isn't one correct way to sort them), which I agree isn't so desirable.

@wswoodruff do you know why the old route config worked despite the order?

Comment (bmleight) [reply lib/index.js]: >  Sorting the routes for users I feel is like a hodge-podge-y thing -- without that happening for you it's pretty easy for routes to not match as **you'd expect**

I'd _expect_ order to matter in these sorts of things, and would add Symphony to Devin's list of examples that function this way.

@wswoodruff, in your PR's description you say: "Switch will render the first route that matches and none else", so it seems like this is geared towards addressing a limitation with the Switch component, or doing something that is more appropriate for that component to do.  If we like adding this type of functionality (or giving a devloper the ability to opt into it), maybe we should explore putting it in a `SortingSwitch` component that you pass a sort function into (someone else may want to sort it differently?).

File: lib/index.js (hunk 2 above)
Comment (wswoodruff) [reply lib/index.js]: Could use some help on alternative strategies for removing duplicate slashes w/out using regex

File: lib/index.js [hunk 7]
```diff
@@ -1,98 +1,128 @@
...
 
-    static propTypes = {
-        history: T.object,
-        routes: T.array.isRequired
-    }
+internals.renderRoutes = (routes) => {
 
-    render() {
+    // Here we're replacing root slash routes (root config route with { ...path: '/' })
+    // with their children because otherwise they will match before anything else
```
Comment (wswoodruff) [inline lib/index.js]: Regarding root slash routes: I can't figure out how to wrap the child components in a component specified. For example:

```
[{
    path: '/',
    component: PlainLayout,
    childRoutes: [
        Home,
        ...Signup,
        ...Login,
        ...Legal
    ]
}]
```

Comment (wswoodruff) [reply lib/index.js]: I can grab ahold of each child, but what do I set the component to on each?

File: lib/index.js (hunk 2 above)
Comment (devinivy) [reply lib/index.js]: I'm imagining just explicitly handling the cases that cause duplicate slashes.  It generally has to do with the base ending with `/` and the route's path beginning with `/`.  Might follow logic similar to this,

```js
const concatPaths = (base, path) => {

    base = base.endsWith('/') ? base.slice(0, -1) : base; // /my-path/ -> /my-path
    path = base.startsWith('/') ? path.slice(1) : path;   // /my-path -> my-path

    return `${base}/${path}`;
};
```

Comment (wswoodruff) [reply lib/index.js]: Thanks! Used that as a helper 👍

File: package.json [hunk 8]
```diff
@@ -33,8 +34,7 @@
     "babel-preset-react": "^6.24.1",
     "babel-preset-stage-0": "^6.24.1"
   },
-  "dependencies": {},
-  "peerDependencies": {
-    "react-router-dom": "^4.2.2"
+  "dependencies": {
+    "react-router-dom": "4.2.x"
```
Comment (devinivy) [inline package.json]: Did we turn this into a regular dependency on purpose?  And any reason we're pinned to this minor version?

Comment (wswoodruff) [reply package.json]: We use `Switch` and `Route` from that lib in lib/index, but I don't think the minor version is important

Comment (wswoodruff) [reply package.json]: Ohh, yeah that's me not understanding peerDependencies again -- yes I agree it's the job of the main project to determine the `react-router-dom` version although this lib requires it to be at least 4.x -- will change back to peerDependencies

File: lib/index.js [hunk 9]
```diff
@@ -1,98 +1,137 @@
...
+exports.buildRoutes = (routes) => internals.renderRoutes(routes);
 
-    static propTypes = {
-        history: T.object,
-        routes: T.array.isRequired
-    }
+internals.renderRoutes = (routes) => {
 
-    render() {
+    // Here we're replacing root slash routes (root config route with { ...path: '/' })
```
Comment (devinivy) [inline lib/index.js]: I forget—was this causing a problem?

Comment (wswoodruff) [reply lib/index.js]: At the root level only -- if you have a path `{ path: '/' }` it will effectively match just like a `{ path: ':someParam' }` and that would cause something the user wouldn't expect. Throughout the rest of the config this is not a problem because there are base parts to each route from then on

Comment (wswoodruff) [reply lib/index.js]: Dealing out the children in place of the parent is a courtesy to the user but much of the value is lost because I haven't figured out how to wrap the children in the slash route's component yet

File: lib/index.js [hunk 10]
```diff
@@ -1,98 +1,137 @@
...
+            rootSlashRoutes[String(i)] = r; // These routes will be keyed by their indices
+        }
+    });
 
-    const renderRouteRecursive = (route, i) => {
+    const toRender = [...routes];
 
-        // These relative paths seem to require a slash in front
+    // Replace slash routes with their children at their indices
+    for (let i = routes.length; i > 1; --i) {
```
Comment (devinivy) [inline lib/index.js]: If there's one route, it seems like this wont run.  Is it supposed to?

I feel a little confused by the `rootSlashRoutes` routine above followed this backwards loop over `routes` that operates on `rootSlashRoutes[i]` (rather than `routes[i]`), that splices `toRender`.  Hummm!!

File: lib/index.js (hunk 9 above)
Comment (devinivy) [reply lib/index.js]: > and that might cause something the user wouldn't expect

Can you give an example?  I don't doubt this is true, but I can't think of an example.

Comment (devinivy) [reply lib/index.js]: (At a glance it seems like they will behave identically.)

Comment (wswoodruff) [reply lib/index.js]: In this config the `:slug` won't match as '/' matches first:

```
[
    // This App(store) config contains routes prefixed by `/app`
    App(store),
    {
        path: '/',
        childRoutes: [
            Home,
            ...Signup,
            ...Login,
            ...Legal
        ]
    },
    {
        path: ':slug',
        component: PlainLayout,
        childRoutes: [
            Signup.specificSchool
        ]
    }
]
```

File: lib/index.js (hunk 10 above)
Comment (wswoodruff) [reply lib/index.js]: Yer right this is all messed up -- I changed this in commit #143f9f2
Still need to go backwards on the array because when splicing things in place gotta go backwards

File: dist/index.js [hunk 11]
```diff
@@ -1,9 +1,5 @@
 'use strict';
 
-var _extends2 = require('babel-runtime/helpers/extends');
```
Comment (devinivy) [inline dist/index.js]: P.S. No pressure to include builds as part of PRs!  Generating a build is a publish-time activity, just like bumping the version, etc.  Not an issue here of course, but it can create issues when merging multiple PRs.

File: lib/index.js (hunk 10 above)
Comment (wswoodruff) [reply lib/index.js]: Yeah I see where ur goin, that sounds good 👍I'd never heard of `.flatten()` before, looking on Mozilla it looks like support is red across the board and the func is experimental. Should we worry about that? Since we're flattening 1 level deep it would be easy enough to replace with `reduce` and `concat` like it says on the Mozilla dev page

Comment (devinivy) [reply lib/index.js]: Yeah, I didn't intend to suggest using `array.flatten()`.  Another one that works is `[].concat(...array)`.

Comment (devinivy) [reply lib/index.js]: Of course it helps to name this stuff, so it's clear what's going on,
```js
const flatten = (arr) => [].concat(...arr);

// ...

flatten(routes);
```

Comment (wswoodruff) [reply lib/index.js]: That blew my mind btw loLL can't believe concat is even cooler than I thought

File: dist/index.js (hunk 11 above)
Comment (wswoodruff) [reply dist/index.js]: The builds keep getting pushed up because I need the build when testing with the branch of Nearpeer I've been using for testing

**Comment by wswoodruff** [comment]
This config will now redirect to 404 if no other routes matched.
Without the redirect setting, the unmatched url will stay in the address bar and any routes that matched will be rendered (The app will probably look broken)
```js
[
    App(store),
    {
        path: '404',
        component: FourOFour
    },
    {
        path: '/',
        component: PlainLayout,
        childRoutes: [
            Home,
            ...Signup,
            ...Login,
            ...Legal
        ],
        redirect: {
            to: '404'
        }
    }
]
```

File: dist/index.js (hunk 11 above)
Comment (devinivy) [reply dist/index.js]: You can always `git add` then `git commit -m`.  You can use modules across projects without pushing using `npm link`.

**Comment by devinivy** [comment]
So are these two identical?
```js
    {
        path: '/',
        component: PlainLayout,
        childRoutes: [
            Home,
            ...Signup,
            ...Login,
            ...Legal
        ],
        redirect: {
            to: '404'
        }
    }
```

and

```js
    {
        path: '/',
        component: PlainLayout,
        childRoutes: [
            Home,
            ...Signup,
            ...Login,
            ...Legal,
            { redirect: {  to: '404' } }
        ]
    }
```

**Comment by wswoodruff** [comment]
Currently that redirect object you have will require a path prop. Then those 2 are identical
`{ path: '', redirect: { to: '404' } }`

**Comment by devinivy** [comment]
The latter is the more general case, and the one most important to support.  From the code I think it will try to render a `<Route/>` component.

**Comment by wswoodruff** [comment]
Ok I'm on it

**Comment by devinivy** [comment]
This example from [the docs](https://reacttraining.com/react-router/web/api/Redirect) seems relevant,
```js
<Switch>
  <Redirect from='/old-path' to='/new-path'/>
  <Route path='/new-path' component={Place}/>
</Switch>
```

Worth keeping in mind that 404 handling is now just a special case of the more general feature of redirecting!

**Comment by wswoodruff** [comment]
Yeah that's really cool that 404 handling is a special case now 👍

**Comment by wswoodruff** [comment]
With commit #e8ec420 you can have a childRoute item that looks how you specified it: `{ redirect: { to: '404' } }`

**Comment by devinivy** [comment]
Awesome!

Can you describe what the `redirect` means here?  Is it really just a way to define a catchall?  If so, is there a reason it shouldn't also be able to be a route?  Should we have two ways of defining a catchall, given that we can already place a catchall at the end of `childRoutes`?  Don't want to overthink this, but also want to keep things simple.  Having one way to do a given thing is a useful constraint.  CC @bmleight for some thoughts.

```js
    {
        path: '/',
        component: PlainLayout,
        childRoutes: [
            Home,
            ...Signup,
            ...Login,
            ...Legal
        ],
        redirect: {
            to: '404'
        }
    }
```

**Comment by wswoodruff** [comment]
The configuration you have there ☝️will add a redirect component to the bottom of the switch component for childRoutes, if childRoutes are listed.

If there are no childRoutes, then this just renders a redirect component, _not_ wrapped in a Switch component.

Since the `component` prop is also optional on routes, you can have configs like this:

```
{
    path: 'oldRoute',
    redirect: {
        to: 'newRoute'
    }
}
```

Question: do we care or even want that? loLL

**Comment by devinivy** [comment]
I would expect that to be,
```
{
    redirect: {
        from: 'oldRoute',
        to: 'newRoute'
    }
}
```

Where `from` is the fully resolved path (when nested deeply).

File: lib/index.js [hunk 12]
```diff
@@ -1,98 +1,139 @@
...
+                                {redirect}
                             </Switch>
-                        }
-                    </route.component>
-                )}
+                        :
+                            redirect;
+
+                    return (
+                        <internals.routeComponentLifecycleWrapper {...props} route={route}>
```
Comment (devinivy) [inline lib/index.js]: If there's no route component, should `routeComponentLifecycleWrapper` wrap `switcher`?

Comment (devinivy) [reply lib/index.js]: Ah, curious!

**Comment by devinivy** [comment]
I don't understand what it would mean to have a `from` on a route's redirect—my impression is that that was specifically for catch-alls.  And I'm not feeling great about redirects not being supported in `childRoutes` if there is a redirect on the route itself.  For me all signs are still pointing to removing the route's `redirect`.  It seems to be causing some issues but not offer any more flexibility.

**Comment by wswoodruff** [comment]
Yeah it _is_ confusing. We should make the `redirect` prop exclusive then and not allow any other props if that one exists

**Comment by devinivy** [comment]
👌 

```js
// This isn't a redirect, or it is a redirect and only has the "redirect" property

!route.redirect || Object.keys(route).length === 1
```

**Comment by wswoodruff** [comment]
With these latest changes, the `onWillMount` that calls `history.replace` is equal in functionality to the redirect below.

```js
{
    path: 'connections',
    // onWillMount: (route, match, location, history) => history.replace('/app/connections/dms'),
    component: ConnectionsTabbedSearch,
    childRoutes: [

        // This redirect may be used in place of the onWillMount func above
        { redirect: { from: '/', exact: true, to: 'dms' } },

        {
            path: 'dms',
            childRoutes: [
                {
                    path: '/',
                    exact: true,
                    component: DirectMessagesPage,
....
```

File: dist/index.js [hunk 13]
```diff
@@ -149,7 +150,7 @@ internals.routeComponentLifecycleWrapper = (_temp = _class = function (_React$Pu
         if (typeof route.componentDidCatch === 'function') {
             _this.componentDidCatch = function (err, info) {
 
-                route.componentDidCatch(err, info, route, match, location, history);
+                route.componentDidCatch({ err: err, info: info, route: route, match: match, location: location, history: history });
```
[the remark this answers is not in the document]
Comment (wswoodruff) [reply dist/index.js]: Oh, that's in the dist =(

[the remark this answers is not in the document]
Comment (wswoodruff) [reply dist/index.js]: I'm sorry I keep pushing that I'm just so used to doing `git add a`

File: lib/index.js [hunk 14]
```diff
@@ -1,98 +1,157 @@
...
+            }
 
-        // These relative paths seem to require a slash in front
+            if (redirect.from) {
+                if (redirect.from === '/') {
 
-        let path = route.path;
-        if (path[0] !== '/') {
-            path = `/${path}`;
+                    redirect.from = basePath;
```
Comment (devinivy) [inline lib/index.js]: There's some mutation of inputs going on in here now.

File: lib/index.js [hunk 15]
```diff
@@ -1,98 +1,157 @@
...
+
+                    // For some reason, redirect.from slash paths do not prepend the entire basePath to redirect's 'to'. It cuts off the last piece so we make it absolute here
+
+                    if (!redirect.to.startsWith(basePath)) {
```
Comment (devinivy) [inline lib/index.js]: No need to check this, I don't think—if it's absolute, don't touch it, and if it's relative then concat the `basePath`.  Is there a reason to do otherwise that you were thinking of?

File: lib/index.js (hunk 14 above)
Comment (wswoodruff) [reply lib/index.js]: I'll revisit that

File: lib/index.js (hunk 15 above)
Comment (wswoodruff) [reply lib/index.js]: I can let it go though

**Comment by devinivy** [comment]
@wswoodruff also keep in mind that `to` can be an object, in which case we probably want to adjust its `pathname`.

File: lib/index.js [hunk 16]
```diff
@@ -1,98 +1,150 @@
...
 
-    const renderRouteRecursive = (route, i) => {
+            if (Object.keys(rest).length !== 0) {
+                throw new Error(`No other properties are allowed alongside "redirect" in route configuration. Check childRoutes of "${basePath}"`);
+            }
+
+            const redirectClone = { ...redirect };
 
-        // These relative paths seem to require a slash in front
+            if (redirectClone.from) {
```
Comment (devinivy) [inline lib/index.js]: `from: ''` could slip through here.

Comment (devinivy) [reply lib/index.js]: Maybe just check if it's a string.

File: lib/index.js [hunk 17]
```diff
@@ -1,98 +1,150 @@
...
-        let path = route.path;
-        if (path[0] !== '/') {
-            path = `/${path}`;
+            if (typeof redirectClone.to === 'string') {
+                // If redirect.to is absolute, leave it be. Otherwise make it relative
+                redirectClone.to = redirectClone.to.startsWith('/') ? redirectClone.to : internals.concatPaths(basePath, redirectClone.to);
+            }
+            else {
+                // to is an object
+                redirectClone.to.pathname = redirectClone.to.pathname.startsWith('/') ? redirectClone.to.pathname : internals.concatPaths(basePath, redirectClone.to.pathname);
```
Comment (devinivy) [inline lib/index.js]: This is mutating `redirect.to`.

File: lib/index.js (hunk 16 above)
Comment (wswoodruff) [reply lib/index.js]: Ok, if we check for a string then `''` and `'/'` will be equivalent. Good?

Comment (devinivy) [reply lib/index.js]: Yeah, that sounds right to me!

File: lib/index.js [hunk 18]
```diff
@@ -42,7 +42,11 @@ internals.renderRoute = (basePath) => {
             }
             else {
                 // to is an object
-                redirectClone.to.pathname = redirectClone.to.pathname.startsWith('/') ? redirectClone.to.pathname : internals.concatPaths(basePath, redirectClone.to.pathname);
+                const pathname = redirectClone.to.pathname.startsWith('/') ? redirectClone.to.pathname : internals.concatPaths(basePath, redirectClone.to.pathname);
```
Comment (devinivy) [inline lib/index.js]: `to` might not have a `pathname`, for example in the case that a 404 is handled just by setting `state`.

Comment (wswoodruff) [reply lib/index.js]: =O oh right! Was used to the path being mandatory
